allennlp/models/coreference_resolution/coref.py

```diff
@@ -0,0 +1,824 @@
+            spans_to_cluster_ids: Dict[Tuple[int, int], int] = {}
+            clusters: List[List[Tuple[int, int]]] = []
+
+            for i, (span, predicted_antecedent) in enumerate(zip(top_spans, predicted_antecedents)):
+                if predicted_antecedent < 0:
+                    # We don't care about spans which are
+                    # not co-referent with anything.
+                    continue
+
+                # Find the right cluster to update with this span.
+                # To do this, we find the row in ``antecedent_indices``
+                # corresponding to this span we are considering.
+                # The predicted antecedent is then an index into this list
+                # of indices, denoting the span from ``top_spans`` which is the
+                # most likely antecedent.
+                predicted_index = antecedent_indices[i, predicted_antecedent]
+
+                antecedent_span = (top_spans[predicted_index, 0],
+                                   top_spans[predicted_index, 1])
+                # Check if we've seen the span before.
+                if antecedent_span in spans_to_cluster_ids.keys():
+                    predicted_cluster_id: int = spans_to_cluster_ids[antecedent_span]
+                else:
+                    # We start a new cluster.
+                    predicted_cluster_id = len(clusters)
+                    # Append a new cluster containing only this span.
+                    clusters.append([antecedent_span])
+                    # Record the new id of this span.
+                    spans_to_cluster_ids[antecedent_span] = predicted_cluster_id
+
+                # Now add the span we are currently considering.
+                span_start, span_end = span
+                clusters[predicted_cluster_id].append((span_start, span_end))
+                spans_to_cluster_ids[(span_start, span_end)] = predicted_cluster_id
+            batch_clusters.append(clusters)
+
+        output_dict["clusters"] = batch_clusters
+        return output_dict
+
+    @overrides
+    def get_metrics(self, reset: bool = False) -> Dict[str, float]:
+        mention_recall = self._mention_recall.get_metric(reset)
+        coref_precision, coref_recall, coref_f1 = self._conll_coref_scores.get_metric(reset)
+
+        return {"coref_precision": coref_precision,
+                "coref_recall": coref_recall,
+                "coref_f1": coref_f1,
+                "mention_recall": mention_recall}
+
+    def _create_attended_span_representations(self,
+                                              head_scores: torch.FloatTensor,
+                                              text_embeddings: torch.FloatTensor,
+                                              span_ends: torch.IntTensor,
+                                              span_width: torch.IntTensor) -> torch.FloatTensor:
+        """
+        Given a tensor of unnormalized attention scores for each word in the document, compute
```

Comment:
I'd recommend "headedness scores" here instead of "attention scores".  It's in several places in this docstring.  It's not an "attention" because there's no query, so the name is surprising.  It's just a score.  Given this, I'd probably also recommend changing the function name to `_create_span_head_representations`.  Seems more intuitive.